File Attachment Support for Activities

infiniterecursion opened the pull request:
Adding logic to attach a file (function Update-WorkItem) to parent work item when the work item is a manual or review activity. If the work item type is a review or manual activity it will attach the file to the parent work item instead of the work item itself as there is no out-of-box way of viewing attachments to activities, and may not have been intended to do so by Microsoft.

Also introducing a logging event for File Attachments within Attach-FileToWorkItem

infiniterecursion committed:
Update smletsExchangeConnector.ps1

Add logic to attach file to parent work item when the work item is a manual or review activity.

AdhocAdam committed:
Work Item guid

Updating Work Item guid property

AdhocAdam committed:
powershell case

updating formatting for smlets

AdhocAdam commented:
For starters - great idea here. Being able to have File Attachments originate from a Review/Manual Activity (e.g. screenshot, some reason for approval/rejection/completion/failure) and push up into the Parent Work Item is something that I never even considered. 👍 

Second - Introducing a File Attachment logging event makes perfect sense/no such event currently exists for that. _But_ if it were to move into the Attach-FileToWorkItem function, logging could be introduced within the file attachment processing loop and thus the event log message could be expanded a bit more to include file specific properties like the file name and/or type. This would also means _every_ current/future call of Attach-FileToWorkItem would trigger the event if logging was set to the required level.

```powershell
$logMessage = "Attaching File
            From: $($message.From)
            Title: $($message.Subject)"
            File Name:
            File Type:
            File Size:
            New-SMEXCOEvent -Source "Attach-FileToWorkItem" -EventId 1 -LogMessage $logMessage -Severity "Information"
```

I'm not committed to _which_ specific file properties should be included in the event, but figured I'd throw out the above example for discussion's sake.

AdhocAdam committed:
Work Item ID

Attach-FileToWorkItem takes the ID (i.e. CR1234) of a Work Item.

infiniterecursion commented:
I like the file properties you have listed, and if I was going to add anything it would be a hash of the file as well. That way if there was some issue with the script pulling the data correctly from EWS you would be able to check it against the original attachment to see that something was getting corrupted. I know it's an expensive operation so a weaker/faster hashing algorithm might be okay as long as we specify in the code that it should not be used for security validation.

AdhocAdam committed:
moving logging call

Moving the logging from Update-WorkItem to Attach-FileToWorkItem. This ensures every file attached is logged and creates the ability to provide more detailed logging around the file itself.

AdhocAdam reviewed (approved):
Looks great!
